src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -61,8 +65,7 @@ protected override string PeerCore
             }
         }
 
-        // TODO(JunTaoLuo, JamesNK): implement this
-        protected override DateTime DeadlineCore => throw new NotImplementedException();
+        protected override DateTime DeadlineCore => _deadline;
```

Comment:
Note that the deadline support means two things:
- exposing the deadline to the user code via server call context
- the framework itself should watch the deadline and terminate the http stream if deadline is reached. Also, the  serverCallContext.CancellationToken should be automatically cancelled by the framework when the deadline is reached (another case when the cancellationToken should be triggered is when the client cancels).

Reply:
I think the right way to terminate a stream on the server side if deadline is reached is to send RST_STREAM (and not to send a status "DeadlineExceeded"), but I will have to confirm that. AFAIK the "DeadlineExceeded" status code is generated on the client side - with gRPC, both client and server are watching the deadline and both parties cancel if it's reached.

Reply:
I've made the CancellationToken use the deadline. Let me know what cancellation should do to the http stream.

Reply:
the server should send RST_STREAM(CANCEL) if the deadline is exceeded.

The processing on client side should look like this (irrelevant now but will be important for managed client):
When the client receives RST_STREAM, it should double-check its local timer to see if it has expired (it's possible it's expired, but not been processed yet). If it has expired, it should throw away the server "response" (the RST) and generate a DEADLINE_EXCEEDED instead.

Reply:
> the server should send RST_STREAM(CANCEL) if the deadline is exceeded.

What about if the client cancels? What does a client cancellation look like?

I've been struggling to find good docs around gRPC cancellation.

Reply:
if client cancels (for whatever reason - it can be that the deadline exceeded on the client side or that the user has decided to cancel), it just sends RST_STREAM(cancel) to notify the server. I'm assuming this is the default behavior when HttpClient decides to cancel an inflight request.